packages/vgrammar-plot/src/semantic-mark.ts

```diff
@@ -0,0 +1,1282 @@
+import type {
+  LegendBaseAttributes,
+  SliderAttributes,
+  DataZoomAttributes,
+  BaseLabelAttrs,
+  PlayerAttributes,
+  OrientType
+} from '@visactor/vrender-components';
+import type { ILogger } from '@visactor/vutils';
+import { Logger, array, isArray, isBoolean, isNil, isPlainObject } from '@visactor/vutils';
+import { isContinuous, type IBaseScale, isDiscrete } from '@visactor/vscale';
+import {
+  getPalette,
+  DiffState,
+  ComponentEnum,
+  invokeFunctionType,
+  getTheme,
+  SIGNAL_VIEW_BOX
+} from '@visactor/vgrammar';
+import type {
+  ISemanticMark,
+  ISemanticMarkSpec,
+  ParsedSimpleEncode,
+  WithDefaultEncode,
+  SemanticTooltipOption,
+  SemanticAxisOption,
+  SemanticPlayerOption,
+  SemanticLabelOption,
+  SemanticDataZoomOption,
+  SemanticSliderOption,
+  SemanticLegendOption,
+  SemanticCrosshairOption,
+  CoordinateOption,
+  PolarCoordinateOption,
+  AxisSpec,
+  CrosshairSpec,
+  DatazoomSpec,
+  DimensionTooltipSpec,
+  LabelSpec,
+  LegendSpec,
+  PlayerSpec,
+  SliderSpec,
+  TooltipSpec,
+  CoordinateSpec,
+  DataSpec,
+  ChannelEncodeType,
+  ScaleSpec,
+  TransformSpec,
+  MarkAnimationSpec,
+  ViewSpec,
+  MarkType,
+  Nil,
+  ValueOf,
+  GenerateBaseEncodeSpec,
+  IElement,
+  IAnimationConfig,
+  MarkSpec,
+  MarkRelativeItemSpec
+} from '@visactor/vgrammar';
+import { field as getFieldAccessor, toPercent } from '@visactor/vgrammar-util';
+import type { IBaseCoordinate } from '@visactor/vgrammar-coordinate';
+import type { ITextAttribute } from '@visactor/vrender';
+
+const defaultTheme = getTheme('default');
+
+let semanticMarkId = -1;
+
+export abstract class SemanticMark<EncodeSpec, K extends string> implements ISemanticMark<EncodeSpec, K> {
+  //declare type: T;
+  spec: ISemanticMarkSpec<EncodeSpec, K>;
+  viewSpec?: ViewSpec;
+
+  private _uid: number;
```

Comment:
Uid is a readonly & public attribute in grammar-base class. Maybe unify the attribute statement here.